### ISSUE TITLE
Do we have any Command line for Gradle as well? like we have it for maven (like using with -Drewrite)

### ISSUE THREAD

**Issue by Sanskarbajaj** [openrewrite/rewrite-docs#257]
Do we have any Command line for Gradle as well? like we have it for maven (like using with -Drewrite)

**Comment by Sanskarbajaj** [comment]
1. firstly just for trying added the dependency in  build.gradle just to orderImports if that works or not . that works fine when i do gradle rewriteRun.

2.  But i want to write custom recipes  which i'm checking the link and trying to replicate is not working. i created a init.gradle file and pasted whatsoever mentioned and created a rewrite.yml file just added one recipe in it which idk is not working i'll share all those to you could you please help me in that :) . And i'm new to Gradle

**Comment by Sanskarbajaj** [comment]
rewrite.yaml
type: specs.openrewrite.org/v1beta/recipe
name: org.openrewrite.FindSpringUses
displayName: Find all Spring uses
description: This is an example of a custom recipe.
recipeList:
  - org.openrewrite.java.spring.boot2.UpgradeSpringBoot_2_7


init.gradle

initscript {
    repositories {
        maven { url "https://plugins.gradle.org/m2" }
    }
    dependencies {
        classpath("org.openrewrite:plugin:latest.release")
    }
}

rootProject {
    plugins.apply(org.openrewrite.gradle.RewritePlugin)
    dependencies {
        rewrite("org.openrewrite:rewrite-java")
    }
    rewrite {
        activeRecipe("org.openrewrite.FindSpringUses")
    }
    afterEvaluate {
        if (repositories.isEmpty()) {
            repositories {
                mavenCentral()
            }
        }
    }
}

**Comment by Sanskarbajaj** [comment]
and one more thing please how i can run the same in shell script ?


for maven what we're doing is like this we have one migration.sh file


mvn $build_options -f "$project_dir" -U "$rewrite_plugin_artifact:run" \
  			-Drewrite.recipeArtifactCoordinates="$rewrite_spring_artifact" \
  			-Drewrite.activeRecipes=UpgradeSpringBoot2_7AndThirdPartyLibraries \
  			-Drewrite.configLocation="$PWD/rewrite.yml"

please help :)

**Comment by timtebeek** [comment]
In your init gradle it seems you're missing a dependency on `rewrite-spring`, since `UpgradeSpringBoot_2_7` is in there:
https://docs.openrewrite.org/recipes/java/spring/boot2/upgradespringboot_2_7
On that documentation page we also provide a sample Gradle init script for you to use or customize
![image](https://github.com/openrewrite/rewrite-docs/assets/1027334/431b59e9-ff2a-4e19-ba03-3fa8063e597f)

Hope that helps!

**Comment by Sanskarbajaj** [comment]
> Cannot add extension with name 'rewrite', as there is an extension already registered with that name.
 still getting this error

**Comment by Sanskarbajaj** [comment]
Build file 'path\build.gradle' line: 5

An exception occurred applying plugin request [id: 'org.openrewrite.rewrite', version: '6.7.0']
> Failed to apply plugin 'org.openrewrite.rewrite'.
   > Cannot add extension with name 'rewrite', as there is an extension already registered with that name.

**Comment by timtebeek** [comment]
If you use `init.gradle` you should remove any `rewrite` plugin or configuration from `build.gradle`; it's either/or, not both. :)

**Comment by timtebeek** [comment]
From your thumbs up I'm guessing that's resolved your issues there @Sanskarbajaj ?

Perhaps we can update the docs to
1. use rewrite-spring as a dependency: that fits in with the example recipe, and includes a lot of the other recipe modules
2. emphasize not to add the plugin to `build.gradle` when using `init.gradle` 

cc @mike-solomon

**Comment by Sanskarbajaj** [comment]
yeah all those issues are gone could you please help me how i can run as a shell script.  i already sent i'm able to run shellscript for maven ones could ypu please help :)


what we're using for maven 

mvn $build_options -f "$project_dir" -U "$rewrite_plugin_artifact:run" \
  			-Drewrite.recipeArtifactCoordinates="$rewrite_spring_artifact" \
  			-Drewrite.activeRecipes=UpgradeSpringBoot2_7AndThirdPartyLibraries \
  			-Drewrite.configLocation="$PWD/rewrite.yml"

**Comment by shanman190** [comment]
On the Gradle side the similar invocation for what you have above would be:

```
# Assuming PWD is the root of the project directory
gradlew rewriteRun -I init.gradle -Drewrite.activeRecipes=org.openrewrite.java.spring.boot2.UpgradeSpringBoot2_7
```

The default config location is the `${rootProject.projectDir}/rewrite.yml`, but it can also be changed by using the initscript. The active recipe can be defined either with the JVM argument above OR it can be defined within the initscript itself as shown in the documentation, but you'd want to pick one or the other to use. Lastly, the recipe id must be it's fully qualified `name`. So as an example, for a recipe defined in `rewrite.yml` this must be the `name` of that recipe as found in that YAML file.

**Comment by shanman190** [comment]
From a docs standpoint, everything that Tim and I have shared is described here:


https://docs.openrewrite.org/running-recipes/running-rewrite-on-a-gradle-project-without-modifying-the-build

https://docs.openrewrite.org/reference/gradle-plugin-configuration


@mike-solomon, unless I missed something it seems like we do miss noting on the Gradle side that you can also control the active recipe and style using JVM arguments. (`activeRecipe`, `activeRecipes`, `rewrite.activeRecipe`, or `rewrite.activeRecipes`; then the same for `activeStyle`, `activeStyles`, `rewrite.activeStyle`, and `rewrite.activeStyles`)

**Comment by mike-solomon** [comment]
I'm going to summarize what I think are the open issues here so it's clear for anyone picking this up (potentially future me). If any of you could confirm/deny/assist with this I'd appreciate it:

- [ ] In the [running rewrite on a gradle project without modifying the build doc](https://docs.openrewrite.org/running-recipes/running-rewrite-on-a-gradle-project-without-modifying-the-build), we have a dependency on `rewrite-java`. That doesn't fit in with the Spring recipe example. Instead, we should add a dependency on `rewrite-spring`.
- [ ] In the same file, we need to clarify that if you use an init script, you **can not** have a `rewrite` section in your `build.gradle` file. 
- [ ] In both the above file, and in the [gradle plugin configuration file](https://docs.openrewrite.org/reference/gradle-plugin-configuration), we should add that as part of running Gradle from the command line that you can add specific JVM args to it. For instance, if you were in the same directory as the `init.gradle` file, you could run the following command to execute the `UpgradeSpringBoot2.7` recipe: `gradlew rewriteRun -I init.gradle -Drewrite.activeRecipes=org.openrewrite.java.spring.boot2.UpgradeSpringBoot2_7`. These arguments include: (`activeRecipe`, `activeRecipes`, `rewrite.activeRecipe`,  `rewrite.activeRecipes`, `activeStyle`, `activeStyles`, `rewrite.activeStyle`, and `rewrite.activeStyles`)

**Comment by timtebeek** [comment]
Yes I agree with those points; for that last one it might even make the most sense to move as much configuration from the `init.gradle` script to command line arguments. That way folks are better able to use a single file and modify just the arguments. A familiar pattern, also as compared to the Maven approach of having mostly commandline arguments.

**Comment by mike-solomon** [comment]
I tried following along with the existing guide and noticed I ran into IntelliJ warnings when making an `init.gradle` script:

![image](https://github.com/openrewrite/rewrite-docs/assets/1187834/d6f67235-dc0f-43af-b277-cd4e363d891b)

Have any of you ran into that before? Maybe not a big deal as I can still run the command in the command line -- but it feels a bit bleh to have other people run into this if they're following along.

Things I've tried include:
* Deleting the `.idea` folder and reopening
* Deleting the same folder and opening from the `build.gradle` file as a project
* Invalidating the cache and restarting

Was looking at: https://stackoverflow.com/questions/32905270/intellij-idea-and-gradle-cannot-be-applied-to-groovy-lang-closure -- but none of those seemed to fix it for me.

**Comment by shanman190** [comment]
So this is something that I see relatively regularly... Intellij's support for Project and Settings scripts is great, but Init just isn't there. This is probably due to it being a fairly uncommon one to work with for the vast majority.

**Comment by mike-solomon** [comment]
> So this is something that I see relatively regularly... Intellij's support for Project and Settings scripts is great, but Init just isn't there. This is probably due to it being a fairly uncommon one to work with for the vast majority.

Fair enough! I just won't worry about it then - thanks!

**Comment by shanman190** [comment]
They are all the same, just various permutations. I believe what happened was that the Gradle plugin and the Maven plugin has diverged, so the @sambsnyd had added the permutations as a mechanism to keep backwards compatibility, but add alignment.

**Comment by mike-solomon** [comment]
PR that fixes all of the above issues: https://github.com/openrewrite/rewrite-docs/pull/260
